### PR TITLE
openssl/quictls support

### PR DESCRIPTION
**Pull request by brbzull0**
Apache Traffic Server uses this library for handling the QUIC side of things of H3, as a part of this effort we plan to use quictls as the  cryptography  library, this is an effort(I work for Yahoo) to support that.

---

This PR includes  all the previous work  done in the [openssl](https://github.com/cloudflare/quiche/tree/openssl) branch.  

### Design Notes

As some of the API are different between vendors, I have added two sub-modules to handle the specifics, for both, tls and the crypto module. The specifics of course are coded inside each submodule:
BoringSSL:
- borinssl_crypto.rs
- borinssl_tls.rs

OpenSSL/quictls:
- openssl_quictls_crypto.rs
- openssl_quictls_tls.rs

Each sub-module will be compiled depending on the feature you use (```openssl``` or ```boringssl vendor```) from the main module(```tls```, ```crypto```).

### Features
```0-RTT``` Is not supported in this PR. It will be added afterwards. This is reflected in the README.

### CI
- we need to work around having openssl/quitls build and let quiche use it.
- pkg-config seems not to be installed as `build-dependency`

### Building notes for testing.

- Make sure you have the openssl library in your `LD_LIBRARY_PATH` and the right path inside the `PKG_CONFIG_PATH`
- Add `openssl` in the cargo `--features` list

**I am using openssl/quictls 3 for this implementation.**

### CODE REVIEW

**Comment by brbzull0** [comment]
CI need to include openssl/quictls so we know we are passing all build/tests with the each vendor.

**Comment by Eimji** [comment]
Hello,

I am working on a mobile application which allows to use encrypted DNS (Android for the moment, iOS right after). 
I tested quiche with a little proto on Android, it works great (thanks!). So I wanted to integrate quiche into my application to develop DoQ and DoH3, but I got compilation errors  due to conflicts with OpenSSL. Yes, I use OpenSSL 3.2 to enable DoT and DoH.

It would be great if we can have a pluggable TLS vendor. 

@brbzull0 As OpenSSL 3.2 supports now QUIC (not all features), your PR will support legacy OpenSSL > 3.2 or only quictls? Thanks for your work.

**Comment by brbzull0** [comment]
> Hello,
> 
> I am working on a mobile application which allows to use encrypted DNS (Android for the moment, iOS right after). I tested quiche with a little proto on Android, it works great (thanks!). So I wanted to integrate quiche into my application to develop DoQ and DoH3, but I got compilation errors due to conflicts with OpenSSL. Yes, I use OpenSSL 3.2 to enable DoT and DoH.
> 
> It would be great if we can have a pluggable TLS vendor.
> 
> @brbzull0 As OpenSSL 3.2 supports now QUIC (not all features), your PR will support legacy OpenSSL > 3.2 or only quictls? Thanks for your work.

Hello @Eimji . The plan is only quictls for now. Thanks for having a look

**Comment by brbzull0** [comment]
Making this ready for review so we can gather some input. Thanks.

**Comment by brbzull0** [comment]
I think is worth having [this](https://github.com/quictls/openssl/issues/138) link here, in case we want to discuss about it.

**Review by ghedo — DISMISSED** [review]

File: quiche/src/crypto/boringssl.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
+use super::*;
+
+use std::mem::MaybeUninit;
+
+use libc::c_int;
+
+// NOTE: This structure is copied from <openssl/aead.h> in order to be able to
+// statically allocate it. While it is not often modified upstream, it needs to
+// be kept in sync.
```
Comment (LPardue) [inline quiche/src/crypto/boringssl.rs:9]: Should we add the commit ID in the comment to make tracking more easy?

File: quiche/src/crypto/mod.rs [hunk 2]
```diff
@@ -274,67 +238,12 @@ impl Seal {
 
             header: HeaderProtectionKey::from_secret(aead, &secret)?,
 
-            packet: PacketKey::from_secret(aead, &secret)?,
+            packet: PacketKey::from_secret(aead, &secret, 1)?,
```
Comment (LPardue) [inline quiche/src/crypto/mod.rs:241]:
```suggestion
            packet: PacketKey::from_secret(aead, &secret, Self::ENCRYPT)?,
```

File: quiche/src/h3/mod.rs [hunk 3]
```diff
@@ -3089,6 +3089,7 @@ mod tests {
         assert!(grease_value() < 2u64.pow(62) - 1);
     }
 
+    #[cfg(not(feature = "openssl"))]
```
Comment (LPardue) [inline quiche/src/h3/mod.rs:3092]:
```suggestion
    #[cfg(not(feature = "openssl"))] // 0-RTT not supported when using openssl/quictls
```

File: quiche/src/lib.rs [hunk 4]
```diff
@@ -14311,7 +14327,9 @@ mod tests {
             })
         );
     }
-
+    // openssl does not provide a straight interface to deal with custom
```
Comment (LPardue) [inline quiche/src/lib.rs:14330]:
```suggestion
    // openssl does not provide a straightforward interface to deal with custom
```

File: quiche/src/tls/openssl_quictls.rs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+pub(super) static QUICHE_STREAM_METHOD: SSL_QUIC_METHOD = SSL_QUIC_METHOD {
+    set_encryption_secrets: Some(set_encryption_secrets),
+    add_handshake_data: Some(super::add_handshake_data),
+    flush_flight: Some(super::flush_flight),
+    send_alert: Some(super::send_alert),
+};
+
+impl Context {
+    pub fn set_early_data_enabled(&mut self, _enabled: bool) {
+        println!("## calling  set_early_data_enabled context!!")
```
Comment (LPardue) [inline quiche/src/tls/openssl_quictls.rs:56]: hangover that needs cleanup. Probably a noop or maybe an assert?

**Review by LPardue — CHANGES_REQUESTED** [review]
A few nits but overall looks good

**Review by ghedo — APPROVED** [review]

**Review by LPardue — APPROVED** [review]

**Comment by ghedo** [comment]
@brbzull0 I was going to do it, just got delayed by travel. Thanks for your work, this is merged now!

**Comment by brbzull0** [comment]
> @brbzull0 I was going to do it, just got delayed by travel. Thanks for your work, this is merged now!

great stuff!
@ghedo @LPardue  Thanks a lot gents!

**Pull request merged**